src/google-classroom-extractor/tests/courses/test_courses.py

```diff
@@ -0,0 +1,128 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+from pathlib import Path
+from unittest.mock import patch
+import pandas as pd
+from sqlalchemy import create_engine
+from google_classroom_extractor.api.courses import request_all_courses_as_df
+
+DB_FILE = "tests/courses/test_courses.db"
+
+
+def dataframe_row_count(dataframe) -> int:
+    return dataframe.shape[0]
+
+
+def db_row_count(test_db) -> int:
+    return test_db.engine.execute("SELECT COUNT(rowid) from Courses").scalar()
+
+
+def db_ending_name(test_db) -> int:
+    return test_db.engine.execute(
+        "SELECT name from Courses WHERE rowid = (SELECT MAX(rowid) from Courses)"
+    ).scalar()
+
+
+def db_room_by_course_id(test_db, course_id) -> int:
+    return test_db.engine.execute(
+        f"SELECT room from Courses WHERE id = '{course_id}'"
+    ).scalar()
+
+
+@patch("google_classroom_extractor.api.courses.request_latest_courses_as_df")
+def test_overlap_removal(mock_latest_courses_df):
+    Path(DB_FILE).unlink(missing_ok=True)
+    test_db = create_engine(f"sqlite:///{DB_FILE}", echo=True)
+
+    # 1st pull: courses-1st.csv is 17 rows ending with "Julia Harding"
+    mock_latest_courses_df.return_value = pd.read_csv("tests/courses/courses-1st.csv")
+    first_courses_df = request_all_courses_as_df(None, test_db)
+    assert dataframe_row_count(first_courses_df) == 17
+    assert db_row_count(test_db) == 17
+    assert db_ending_name(test_db) == "Julia Harding"
+
+    # 2nd pull: courses-2nd-overlaps-1st is 59 rows with 5 overlapping ending with "Maurice Hatfield"
+    mock_latest_courses_df.return_value = pd.read_csv(
+        "tests/courses/courses-2nd-overlaps-1st.csv"
+    )
+    second_courses_df = request_all_courses_as_df(None, test_db)
+    assert dataframe_row_count(second_courses_df) == 59
+    assert db_row_count(test_db) == 71  # 59 new + 17 existing - 5 overlapping
+    assert db_ending_name(test_db) == "Maurice Hatfield"
```

Comment:
Is there any reasonable way to split this into multiple tests, so that if, for example, line 45 fails, we'll still have a chance to evaluate the block of asserts immediately following line 51?

Reply:
they are together because they are dependent.  i'm testing behavior across multiple pulls and asserting correctness after each.

Reply:
OK